lib/services/nodes-api-service.js

```diff
@@ -30,64 +32,136 @@ function nodeApiServiceFactory(
-     * @param  {Object}     node
-     * @param  {Array}      enclNodes
-     * @return {Promise}
+     * Remove the relations to node in the downstram node. If the relation field is
+     * empty after removing, delete this field.
+     * @param  {Object}     node node whose relation needs to be updated
+     * @param  {String}     type relation type that needs to be updated
+     * @param  {String}     targetId id in relation type that needs to be deleted
+     * @return {Promise}    node after removing relation
      */
-    NodeApiService.prototype._removeEnclRelations = function(node, enclNodes) {
-        if (!enclNodes) {
+    NodeApiService.prototype._removeRelation = function(node, type, targetId) {
+
+        if (!node) {
             return Promise.resolve();
         }
 
-        // Remove the relationship between enclosure and compute node
-        return Promise.map(enclNodes, function(enclNode) {
-            if (!_.has(enclNode, 'relations')) {
-                return;
-            }
+        // Remove the relationship between node and downstram node
+        if (!_.has(node, 'relations')) {
+            return Promise.resolve();
+        }
 
-            var index = _.findIndex(enclNode.relations, { relationType: 'encloses' });
-            if (index === -1 || !_.has(enclNode.relations[index], 'targets')) {
-                return;
-            }
+        var index = _.findIndex(node.relations, { relationType: type });
+        if (index === -1 || !_.has(node.relations[index], 'targets')) {
+            return Promise.resolve();
+        }
+
+        // Remove target node id in relation field
+        if (_.indexOf(node.relations[index].targets, targetId) !== -1) {
+            _.pull(node.relations[index].targets, targetId);
+        }
+
+        // Remove the type of relation if no targets in it
+        if (node.relations[index].targets.length === 0) {
+            _.pull(node.relations, node.relations[index]);
+        }
+
+        return waterline.nodes.updateByIdentifier(node.id,
+                                                  {relations: node.relations});
+    };
+
+    /**
+     * Handle target nodes. Update relation type field and remove the target
+     * node if needed.
+     * @param  {Object}     targetNodos array of target nodes
+     * @param  {Object}     type relation type in the removing node
+     * @param  {Object}     nodeId removing node id
+     * @return {Promise}
+     */
+    NodeApiService.prototype._handleTargetNodes = function(targetNodes, type, nodeId) {
+        var self = this;
+        var targetType = '';
+        var relationClass = null;
+        var needDel = false;
+
+        if (Constants.NodeRelations.hasOwnProperty(type)) {
+            targetType = Constants.NodeRelations[type].mapping;
+            relationClass = Constants.NodeRelations[type].relationClass;
+        }
+
+        if ((relationClass === 'component') &&
```

Comment:
this would probably make more sense to include this logic (to delete or not) within the promise chain (around line 125) since you're starting with a single node in this function and cascading to build a map of all nodes you want to delete. If you're expecting an input of nodes, then you may want to construct a promise chain from that input (assuming the input is an array of nodes) and utilize Promise.each.